File: FastAPI/app/mods/connectionManager.py
import json
from typing import List, Dict, Set
from fastapi import WebSocket
import json
from datetime import datetime
import time
import redis
import threading
import asyncio


class ConnectionManager:
    __slots__ = "id", "sessions", "red", "sub", "thread"

    # TEST #

    def test_set_get(self):
        self.red.set("test", "test")
        test = self.red.get("test")
        return test

    # ENDOFTESTTS #

    def __init__(self, redis_host, redis_port):
        self.id = str(time.time())
        self.sessions: Dict = {}
        # check redis connection refused and catch it later
        self.red = redis.Redis(
            connection_pool=redis.ConnectionPool(
                host=redis_host, port=redis_port, db=0, max_connections=200
            ),
            charset="utf-8",
            decode_responses=True,
        )
        self.sub = self.red.pubsub()
        self.sub.subscribe([self.id])
        self.thread = threading.Thread(target=self.between_callback)
        self.thread.daemon = True
        self.thread.start()

    def between_callback(self):
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        loop.run_until_complete(self.listen_to_publisher())
        loop.close()

    async def listen_to_publisher(self):
        for item in self.sub.listen():
            if item["data"] != 1:
                data = json.loads(item["data"])  # needs exception
                command = data["command"]
                if command is not None:
                    command_string = command["command"]
                    if command_string == "add_to_subscribers":
                        username_target = command["username_target"]
                        subscriber = command["subscriber"]
                        if username_target in self.sessions:
                            self.sessions[username_target]["subscribers"].add(
                                subscriber
                            )
                else:
                    username_target = data["username_target"]
                    message_dict = data["message_dict"]
                    if username_target in self.sessions:
                        for socket in self.sessions[username_target]["sockets"]:
                            await socket.send_text(json.dumps(message_dict))

    def notify(self, username_target, message_dict, command=None):
        server_ids = self.red.smembers(username_target)
        for server_id in server_ids:
            if server_id.decode("utf-8") != self.id:
                self.red.publish(
                    server_id,
                    json.dumps(
                        {
                            "username_target": username_target,
                            "message_dict": message_dict,
                            "command": command,
                        }
                    ),
                )

    async def connect(self, username: str, websocket: WebSocket, subscribers: Set[str]):
        await websocket.accept()
        if username not in self.sessions:
            self.sessions[username] = {
                "sockets": [websocket],
                "subscribers": subscribers,
            }
            # add server to user
            self.red.sadd(username, self.id)
            # notify subscribers
            message_dict = {
                "event_type": "status",
                "username_target": username,
                "status": "online",
            }
            for subscriber in subscribers:
                if subscriber in self.sessions:
                    for socket in self.sessions[subscriber]["sockets"]:
                        await socket.send_text(json.dumps(message_dict))
                self.notify(subscriber, message_dict)
        else:
            self.sessions[username]["sockets"].append(websocket)
        print(self.sessions)

    async def disconnect(self, username: str, websocket: WebSocket):
        if username in self.sessions:
            self.sessions[username]["sockets"].remove(websocket)
            if len(self.sessions[username]["sockets"]) < 1:
                # remove server from user
                self.red.srem(username, self.id)
                # check if user is offline
                servers = self.red.smembers(username)
                if len(servers) < 1:
                    message_dict = {
                        "event_type": "status",
                        "username_target": username,
                        "status": "offline",
                    }
                    for subscriber in self.sessions[username]["subscribers"]:
                        if subscriber in self.sessions:
                            for socket in self.sessions[subscriber]["sockets"]:
                                await socket.send_text(json.dumps(message_dict))
                        self.notify(subscriber, message_dict)
                del self.sessions[username]

    def add_to_subscribers(self, subscriber, username_to):
        if username_to in self.sessions:
            self.sessions[username_to]["subscribers"].add(subscriber)
        command = {
            "command": "add_to_subscribers",
            "subscriber": subscriber,
            "username_target": username_to,
        }
        self.notify(username_to, None, command)

    async def request_status(self, subscriber, usernames_to):
        for username_to in usernames_to:
            serves = self.red.smembers(username_to)
            if len(serves) > 0:
                message_dict = {
                    "event_type": "status",
                    "username_target": username_to,
                    "status": "online",
                }
                if subscriber in self.sessions:
                    for socket in self.sessions[subscriber]["sockets"]:
                        await socket.send_text(json.dumps(message_dict))
                self.notify(subscriber, message_dict)

    async def send_personal_typing(self, username_from: str, username_to: str):
        message_dict = {"event_type": "typing", "username_target": username_from}
        if username_to in self.sessions:
            for socket in self.sessions[username_to]["sockets"]:
                await socket.send_text(json.dumps(message_dict))
        self.notify(username_to, message_dict)

    async def send_personal_message(
        self, username_from: str, username_to: str, text_content: str
    ):
        date = datetime.now().strftime("%H:%M:%S")
        message_id = time.time()
        message_dict = {
            "event_type": "msg",
            "event_content": {
                "username_target": username_to,
                "message_content": {
                    "type": "txt",
                    "text_content": text_content,
                    "date": date,
                    "id": message_id,
                    "received": False,
                },
            },
        }
        if username_from in self.sessions:
            for socket in self.sessions[username_from]["sockets"]:
                await socket.send_text(json.dumps(message_dict))
        self.notify(username_from, message_dict)
        message_dict = {
            "event_type": "msg",
            "event_content": {
                "username_target": username_from,
                "message_content": {
                    "type": "txt",
                    "text_content": text_content,
                    "date": date,
                    "id": message_id,
                    "received": True,
                },
            },
        }
        if username_to in self.sessions:
            for socket in self.sessions[username_to]["sockets"]:
                await socket.send_text(json.dumps(message_dict))
        self.notify(username_to, message_dict)

    async def send_personal_like(
        self, username_from: str, username_to: str, message_id: str
    ):
        message_dict = {
            "event_type": "like",
            "username_target": username_to,
            "message_id": message_id,
            "self_like": True,
        }
        if username_from in self.sessions:
            for socket in self.sessions[username_from]["sockets"]:
                await socket.send_text(json.dumps(message_dict))
        self.notify(username_from, message_dict)
        message_dict = {
            "event_type": "like",
            "username_target": username_from,
            "message_id": message_id,
        }
        if username_to in self.sessions:
            for socket in self.sessions[username_to]["sockets"]:
                await socket.send_text(json.dumps(message_dict))
        self.notify(username_to, message_dict)

File: ExpressAPI/mods/connectionManager.js
const Redis = require('redis');


function removeFromArray(array, value) {
  return array.filter((element) => {
    return element != value;
  });
}


class ConnectionManager {
  constructor(redisHost, redisPort) {
    this.id = (Date.now() / 1000).toString();
    this.sessions = {};
    this.red = Redis.createClient({
      url: `redis://${redisHost}:${redisPort}`
    });
    this.sub = this.red.duplicate();
    this.red.connect();
    this.sub.connect();
    this.sub.subscribe('events', (message) => {
      this.listenToPublisher(message);
    });
  }

  //TESTS
  async testSetGet() {
    await this.red.set('test', 'test');
    let test = await this.red.get('test');
    return test;
  }
  //ENDOFTESTS

  listenToPublisher(message) {
    const data = JSON.parse(message);
    const command = data.command;
    if (command != null) {
      const commandString = command.command;
      if (commandString === "add_to:subscribers") {
        const usernameTarget = command.username_target;
        const subscriber = command.subscriber;
        if (this.sessions.hasOwnProperty(usernameTarget)) {
          this.sessions[usernameTarget]["subscribers"].push(subscriber)
        }
      }
    } else {
      const usernameTarget = data.username_target;
      const messageObj = data.message_dict;
      if (this.sessions.hasOwnProperty(usernameTarget)) {
        for (var i = 0; i < this.sessions[usernameTarget]["sockets"].length; i++) {
          this.sessions[usernameTo]["sockets"][i].send(JSON.stringify(messageObj))
        }
      }
    }
  }


  async notify(usernameTarget, messageObj, command) {
    const serverIds = await this.red.sMembers(usernameTarget);
    for (var i = 0; i < serverIds.length; i++) {
      this.red.publish(serverIds[i], JSON.stringify({ username_target: usernameTarget, message_dict: messageObj, command: command }));
    }
  }


  async connect(username, websocket, subscribers) {
    if (!this.sessions.hasOwnProperty(username)) {
      this.sessions[username] = { sockets: [websocket], subscribers: subscribers };
      // add server to user
      await this.red.sAdd(username, this.id);
      // notify subs
      const messageObj = { event_type: "status", username_target: username, status: "online" };
      for (var i = 0; i < subscribers.length; i++) {
        const subscriber = subscribers[i];
        if (this.sessions.hasOwnProperty(subscriber)) {
          for (var j = 0; j < this.sessions[subscriber]["sockets"].length; j++) {
            this.sessions[subscriber]["sockets"][j].send(JSON.stringify(messageObj))
          }
        }
      }
    } else {
      this.sessions[username]["sockets"].push(websocket);
    }
    console.log(this.sessions)
  }


  async disconnect(username, websocket) {
    if (this.sessions.hasOwnProperty(username)) {
      this.sessions[username]["sockets"] = removeFromArray(this.sessions[username]["sockets"], websocket);
      if (this.sessions[username]["sockets"].length < 1) {
        // remove serve from user
        await this.red.sRem(username, this.id);
        // check if user is offline
        const servers = await this.red.sMembers(username);
        if (servers.length < 1) {
          const messageObj = { event_type: "status", username_target: username, status: "offline" };
          const subscribers = this.sessions[username]["subscribers"];
          for (var i = 0; i < subscribers.length; i++) {
            const subscriber = subscribers[i];
            if (this.sessions.hasOwnProperty(subscriber)) {
              for (var j = 0; j < this.sessions[subscriber]["sockets"].length; j++) {
                this.sessions[subscriber]["sockets"][j].send(JSON.stringify(messageObj));
                console.log("offline sent")
              }
            }
            await this.notify(subscriber, messageObj);
          }
        }
        delete this.sessions[username];
      }
    }
  }


  async addToSubscribers(subscriber, usernameTo) {
    if (!this.sessions.hasOwnProperty(usernameTo)) {
      this.sessions[usernameTo]["subscribers"].push(subscriber);
    }
    const command = { command: "add_to_subscribers", subscriber: subscriber, username_target: usernameTo };
    await this.notify(usernameTo, null, command);
  }


  async requestStatus(subscriber, usernamesTo) {
    for (var i = 0; i < usernamesTo.length; i++) {
      const usernameTo = usernamesTo[i];
      const servers = await this.red.sMembers(usernameTo);
      if (servers.length > 0) {
        const messageObj = { event_type: "status", username_target: usernameTo, status: "online" };
        if (this.sessions.hasOwnProperty(subscriber)) {
          for (var j = 0; j < this.sessions[subscriber]["sockets"].length; j++) {
            this.sessions[subscriber]["sockets"][j].send(JSON.stringify(messageObj));
          }
        }
        await this.notify(subscriber, messageObj);
      }
    }
  }

  async sendPersonalTyping(usernameFrom, usernameTo) {
    const messageObj = { event_type: "typing", "username_target": usernameFrom };
    if (this.sessions.hasOwnProperty(usernameTo)) {
      for (var i = 0; i < this.sessions[usernameTo]["sockets"].length; i++) {
        this.sessions[usernameTo]["sockets"][i].send(JSON.stringify(messageObj))
      }
    }
    await this.notify(usernameTo, messageObj);
  }


  async sendPersonalMessage(usernameFrom, usernameTo, textContent) {
    const today = new Date();
    const date = today.getHours() + ":" + today.getMinutes() + ":" + today.getSeconds();
    const messageId = Date.now() / 1000;
    let messageObj = { event_type: "msg", event_content: { username_target: usernameTo, message_content: { type: "txt", text_content: textContent, date: date, id: messageId, received: false } } };
    if (this.sessions.hasOwnProperty(usernameFrom)) {
      for (var i = 0; i < this.sessions[usernameFrom]["sockets"].length; i++) {
        this.sessions[usernameFrom]["sockets"][i].send(JSON.stringify(messageObj))
      }
    }
    await this.notify(usernameFrom, messageObj);
    messageObj = { event_type: "msg", event_content: { username_target: usernameFrom, message_content: { type: "txt", text_content: textContent, date: date, id: messageId, received: true } } }
    if (this.sessions.hasOwnProperty(usernameTo)) {
      for (var i = 0; i < this.sessions[usernameTo]["sockets"].length; i++) {
        this.sessions[usernameTo]["sockets"][i].send(JSON.stringify(messageObj))
      }
    }
    await this.notify(usernameTo, messageObj);
  }


  async sendPersonalLike(usernameFrom, usernameTo, messageId) {
    let messageObj = { event_type: "like", username_target: usernameTo, message_id: messageId, self_like: true };
    if (this.sessions.hasOwnProperty(usernameFrom)) {
      for (var i = 0; i < this.sessions[usernameFrom]["sockets"].length; i++) {
        this.sessions[usernameFrom]["sockets"][i].send(JSON.stringify(messageObj))
      }
    }
    await this.notify(usernameFrom, messageObj);
    messageObj = { event_type: "like", username_target: usernameFrom, message_id: messageId };
    if (this.sessions.hasOwnProperty(usernameTo)) {
      for (var i = 0; i < this.sessions[usernameTo]["sockets"].length; i++) {
        this.sessions[usernameTo]["sockets"][i].send(JSON.stringify(messageObj))
      }
    }
    await this.notify(usernameTo, messageObj);
  }

}

module.exports = ConnectionManager;